feat: P4ADEV-3876 implemented new fields

dan-ntt opened the pull request:
#### Description
<!--- Please always add a PR description as if nobody knows anything about the context these changes come from. -->
<!--- Even if we are all from our internal team, we may not be on the same page. -->
<!--- Write this PR as you were contributing to a public OSS project, where nobody knows you and you have to earn their trust. -->
<!--- This will improve our projects in the long run! Thanks. -->

#### List of Changes
<!--- Describe your changes in detail -->

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- Pre-Deploy Test
  - [x] Unit
  - [ ] Integration (Narrow)
- Post-Deploy Test
  - [ ] Isolated Microservice
  - [ ] Broader Integration
  - [ ] Acceptance
  - [ ] Performance & Load

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] PATCH - Bug fix (backwards compatible bug fixes)
- [x] MINOR - New feature (add functionality in a backwards compatible manner)
- [ ] MAJOR - Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] CHORE - Minor Change (fix or feature that don't impact the functionality e.g. Documentation or lint configuration)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

Commit:
feat: P4ADEV-3876 implemented new fields

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/model/AssessmentsDetail.java:
```diff
@@ -43,4 +44,12 @@ public class AssessmentsDetail extends BaseEntity implements Serializable {
+  private String officeDescription;
+  private String sectionDescription;
+  private String assessmentDescription;
+  private Long debtPositionTypeOrgId;
```
move this before `debtPositionTypeOrgCode` field
```suggestion
  @NotNull
  private Long debtPositionTypeOrgId;
```

antonioT90 replied:
This has not been resolved

antonioT90 replied:
not resolved yet

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/model/AssessmentsDetail.java:
```diff
@@ -43,4 +44,12 @@ public class AssessmentsDetail extends BaseEntity implements Serializable {
   private boolean amountSubmitted = true;
   private Long receiptId;
+  private String officeDescription;
+  private String sectionDescription;
```
move these after the related `*Code` field

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/service/assessments/AssessmentsDetailServiceImpl.java:
```diff
@@ -114,6 +136,14 @@ private static AssessmentsDetail buildAssessmentsDetail(ReceiptNoPII receipt, In
+            .classificationLabel(ClassificationLabel.REPORTED) // @TODO: logic behind this value to be defined
+            .dateReceipt(receipt.getCreationDate())
+            .dateReporting(installment.getNotificationDate())
+            .dateTreasury(receipt.getTransferDate())
```
these fields will be filled by AssessmentClassificationWF, let's remove them here
```suggestion
```

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/model/AssessmentsDetail.java:
Have we already merged and applied the PR on `p4pa-db` to add these fields?

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/service/assessments/AssessmentsDetailServiceImpl.java:
```diff
@@ -69,7 +70,7 @@ public AssessmentsDetailServiceImpl(AssessmentsDetailRepository assessmentsDetai
   @Override
   public void createAssessmentDetail(Assessments assessments, ReceiptNoPII receipt, InstallmentNoPII installmentNoPII) {
-    List<AssessmentsDetail> assessmentsDetailList = buildAssessmentDetail(receipt, installmentNoPII, assessments);
+    List<AssessmentsDetail> assessmentsDetailList = buildAssessmentDetail(receipt, installmentNoPII, assessments, null, null, null, null);
```
why null values? `buildAssessmentDetail` method could retrieve theme by itself! it's not required to have them as input

antonioT90 commented on openapi/p4pa-classification.openapi.yaml:
```diff
@@ -1460,3 +1460,12 @@ components:
+          type: string
+        debtPositionTypeOrgId:
+          type: integer
+          format: int64
```
these are not API input, they are retrieved from the DB

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/service/assessments/AssessmentsDetailServiceImpl.java:
```diff
@@ -114,6 +136,14 @@ private static AssessmentsDetail buildAssessmentsDetail(ReceiptNoPII receipt, In
+            .officeDescription(officeDescription)
+            .sectionDescription(sectionDescription)
+            .assessmentDescription(assessmentDescription)
+            .debtPositionTypeOrgId(debtPositionTypeOrgId)
```
@robertopasetti what about storing this info also on `Assessments` entity? this way, here we should just copy the field from the input `assessment`

antonioT90 replied:
ps: @dan-ntt let's move it before debtPositionTypeOrgCode setting

robertopasetti replied:
why should we store at Assessment level? Each assessment detail could have its own office, section, assessment data. Instead it could be useful for debt position type org

antonioT90 replied:
@robertopasetti I'm talking just about the `debtPositionTypeOrgId`

robertopasetti replied:
ok, it's a good idea. what about to store it both assessment and assessment_detail level?

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/service/assessments/AssessmentsDetailServiceImpl.java:
```diff
@@ -114,6 +136,14 @@ private static AssessmentsDetail buildAssessmentsDetail(ReceiptNoPII receipt, In
             .assessmentCode(assessmentCode)
             .amountCents(amountCents)
             .receiptId(receipt.getReceiptId())
+            .officeDescription(officeDescription)
+            .sectionDescription(sectionDescription)
+            .assessmentDescription(assessmentDescription)
```
these should be read from `AssessmentRegistry` entity!
* On automatic creation flow (byReceipt) you should search it before to call this build method
** @robertopasetti how should we search it? Filtering by `organizationId`, `debtPositionTypeOrgCode`, `sectionCode`, `officeCode`, `assessmentCode`, `operatingYear` (how should this be eventually configured?)? Should we also filter by `status`?
* On manual creation you should already have ready an `AssessmentRegistry` record to use.

robertopasetti replied:
about automatic flow: we should filter by `organizationId`, `debtPositionTypeOrgCode`, `sectionCode`, `officeCode`, `assessmentCode`, `operatingYear` ignoring `status`

antonioT90 replied:
@robertopasetti `operatingYear` is the current year? are we sure about to ignore the `status`? what if there are many record? can we just take the first?

robertopasetti replied:
the `operatingYear` is the payment date year, about `status` could we have (for the same debt position org and operating year) different values for `assessment` `office` and `section`?

antonioT90 replied:
there is a UNIQUE but only when the status is `ACTIVE`, there could be many having different statuses

robertopasetti replied:
so, we have to search for record in the `ACTIVE` status, if there's no record take the first `INACTIVE` / `CLOSED`

Commit:
feat: some changes based on PR comments

Commit:
Merge branch 'develop' into P4ADEV-3876-new-fields-office

# Conflicts:
#	openapi/generated.openapi.json

Commit:
build: regenerated OpenAPI specs after merge

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/connector/debtposition/client/DebtPositionTypeOrgClient.java:
```diff
@@ -39,4 +39,9 @@ public DebtPositionTypeOrg getDebtPositionTypeOrgByDebtPositionTypeOrgId(Long or
   }
+
+  public DebtPositionTypeOrg getDebtPositionTypeOrgByDebtPositionTypeOrgCode(Long organizationId, String debtPositionTypeOrgCode, String accessToken) {
+    return debtPositionApisHolder.getDebtPositionTypeOrgSearchControllerApi(accessToken)
```
let's handle 404 to return null instead

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/model/Assessments.java:
```diff
@@ -34,4 +34,5 @@ public class Assessments extends BaseEntity implements Serializable {
   private boolean flagManualGeneration;
   @NotNull
   private String operatorExternalUserId;
+  private Long debtPositionTypeOrgId;
```
@NotNull, move before debtPositionTypeOrgCode.
Have you add this new column on p4pa-db PR?

dan-ntt replied:
not yet, I will update the p4pa-db once this changes have been approved

antonioT90 replied:
let's first merge p4pa-db instead! This PR depends on it!

antonioT90 replied:
> @NotNull, move before debtPositionTypeOrgCode.

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/repository/AssessmentsRegistryRepository.java:
```diff
@@ -98,4 +99,24 @@ void insertIfNotExists(
+            AND entity.assessmentCode = :assessmentCode
+            AND entity.operatingYear = :operatingYear
+    """)
+    Optional<AssessmentsRegistry> findByOrganizationIdAndCodes(
```
This could return more than 1 entity, you should take, if exists, the record having ACTIVE status, or any record if there is not ACTIVE
```suggestion
    List<AssessmentsRegistry> findByOrganizationIdAndCodes(
```

Commit:
feat: aligned properties in the entity

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/service/assessments/AssessmentsDetailServiceImpl.java:
```diff
@@ -92,14 +97,55 @@ List<AssessmentsDetail> buildAssessmentDetail(ReceiptNoPII receipt, InstallmentN
+    String sectionCode,
+    String assessmentCode,
+    Long amountCents,
+    Long assessmentsRegistryId
```
make the overloaded method accept the AssessmentRegistry, not it's id...
the version of the method without this parameter would perform the search and call this one

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/service/assessments/AssessmentsDetailServiceImpl.java:
```diff
@@ -114,6 +160,10 @@ private static AssessmentsDetail buildAssessmentsDetail(ReceiptNoPII receipt, In
+            .officeDescription(assessmentsRegistry.map(AssessmentsRegistry::getOfficeDescription).orElse(null))
+            .sectionDescription(assessmentsRegistry.map(AssessmentsRegistry::getSectionDescription).orElse(null))
+            .assessmentDescription(assessmentsRegistry.map(AssessmentsRegistry::getAssessmentDescription).orElse(null))
+            .debtPositionTypeOrgId(assessment.getDebtPositionTypeOrgId())
```
let's order these fields as done in the Entity (dpTOid before the code, descriptions after codes

Commit:
feat: improved registry lookup and enhanced descriptions handling

- Added ordering by active status in registry query.
- Ensured proper resolution of registry and fallback logic.
- Improved null-safety for descriptions.
- Added handling for `HttpClientErrorException.NotFound` in DebtPositionTypeOrg lookup.

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/repository/AssessmentsRegistryRepository.java:
```diff
@@ -98,4 +99,25 @@ void insertIfNotExists(
+            AND entity.officeCode = :officeCode
+            AND entity.assessmentCode = :assessmentCode
+            AND entity.operatingYear = :operatingYear
+        ORDER BY CASE WHEN entity.status = 'ACTIVE' THEN 0 ELSE 1 END ASC
```
ok, you should still configure the logic to take just the first record... asking for an Optional would throw an exception if the query fetch more than 1 record

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/service/assessments/AssessmentsDetailServiceImpl.java:
```diff
@@ -92,26 +97,82 @@ List<AssessmentsDetail> buildAssessmentDetail(ReceiptNoPII receipt, InstallmentN
+    String assessmentCode,
+    Long amountCents
+  ) {
+    return this.buildAssessmentsDetail(receipt, installment, assessment, officeCode, sectionCode, assessmentCode, amountCents, null);
```
fetch here the AssessmentsRegistry instead of check for null on the overloaded method

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/model/AssessmentsDetail.java:
```diff
@@ -23,6 +24,7 @@ public class AssessmentsDetail extends BaseEntity implements Serializable {
   private Long assessmentId;
   @NotNull
   private Long organizationId;
+  private Long debtPositionTypeOrgId;
```
```suggestion
  @NotNull
  private Long debtPositionTypeOrgId;
```

Commit:
feat: refined registry retrieval and error handling

- Moved registry lookup to main logic, ensuring early resolution.
- Updated exception handling for missing DebtPositionTypeOrg.
- Adjusted descriptions mapping for clarity.
- Added query limit for registry search.

Commit:
feat: updated the openapi specs

antonioT90 commented on src/main/java/it/gov/pagopa/pu/classification/service/assessments/AssessmentsDetailServiceImpl.java:
```diff
@@ -92,26 +97,77 @@ List<AssessmentsDetail> buildAssessmentDetail(ReceiptNoPII receipt, InstallmentN
+      sectionCode,
+      officeCode,
+      assessmentCode,
+      String.valueOf(LocalDateTime.now().getYear())
```
@robertopasetti should be used receipt `paymentDateTime` year?

robertopasetti replied:
yes

Commit:
feat: added new test for lookup by code and updated OpenAPI specs

- Introduced test to validate `getDebtPositionTypeOrgByDebtPositionTypeOrgCode` behavior.
- Added `debtPositionTypeOrgId` to `AssessmentsDetail` entity.
- Updated and aligned OpenAPI specs with new models and endpoints.

Commit:
fix: handle null paymentDateTime gracefully in year resolution

- Updated year calculation to use `paymentDateTime` if available, defaulting to current year with `OffsetDateTime.now()` when null.

antonioT90 reviewed (changes requested):
let's resolve Sonar issues

Commit:
fix: removed all unused imports everywhere

Commit:
fix: changed parameter numbers to a DTO

Commit:
Merge branch 'develop' into P4ADEV-3876-new-fields-office

# Conflicts:
#	openapi/generated.openapi.json

Commit:
fix: openapi regeneration
